File: src/types/persistent-tunnel.ts
export type PersistentTunnelProvider = "cloudflared";
export type PersistentTunnelStatus = "stopped" | "starting" | "running" | "error";

export interface PersistentTunnelSetupStatus {
  provider: PersistentTunnelProvider;
  ready: boolean;
  managed: boolean;
  binaryPath?: string | null;
  authCertPath?: string | null;
  credentialsPath?: string | null;
  tunnelId?: string | null;
  tunnelName?: string | null;
  defaultHostnameZone?: string | null;
  details: string;
  guidance?: string | null;
}

export interface PersistentTunnelNamedTunnelSummary {
  tunnelId: string;
  tunnelName: string;
  credentialsPath?: string | null;
  selected: boolean;
}

export interface CreatePersistentNamedTunnelInput {
  name: string;
}

export interface SelectPersistentNamedTunnelInput {
  tunnelId: string;
}

export interface UpdatePersistentTunnelSetupInput {
  defaultHostnameZone?: string | null;
}

export interface ProjectPersistentHostname {
  id: string;
  projectId: string;
  provider: PersistentTunnelProvider;
  hostname: string;
  createdAt: string;
  updatedAt: string;
}

export interface UpsertProjectPersistentHostnameInput {
  projectId: string;
  hostname: string;
}

export interface ApplyProjectPersistentHostnameInput {
  projectId: string;
  hostname?: string | null;
}

export interface ProjectPersistentTunnelState {
  projectId: string;
  provider: PersistentTunnelProvider;
  status: PersistentTunnelStatus;
  hostname: string;
  localUrl: string;
  publicUrl: string;
  logPath: string;
  binaryPath?: string | null;
  tunnelId?: string | null;
  credentialsPath?: string | null;
  updatedAt: string;
  details?: string | null;
}

export interface ApplyProjectPersistentHostnameResult {
  hostname: ProjectPersistentHostname;
  tunnel: ProjectPersistentTunnelState;
}

export interface DeleteProjectPersistentHostnameResult {
  hostname: string;
}

export interface PersistentTunnelHealthCheck {
  code: string;
  label: string;
  status: PersistentTunnelStatus;
  message: string;
}

export interface PersistentTunnelHealthReport {
  projectId: string;
  hostname?: string | null;
  overallStatus: PersistentTunnelStatus;
  checks: PersistentTunnelHealthCheck[];
  updatedAt: string;
}

File: src-tauri/src/models/persistent_tunnel.rs
use serde::{Deserialize, Serialize};
use std::str::FromStr;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PersistentTunnelProvider {
    Cloudflared,
}

impl PersistentTunnelProvider {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Cloudflared => "cloudflared",
        }
    }
}

impl FromStr for PersistentTunnelProvider {
    type Err = &'static str;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "cloudflared" => Ok(Self::Cloudflared),
            _ => Err("Invalid persistent tunnel provider"),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectPersistentHostname {
    pub id: String,
    pub project_id: String,
    pub provider: PersistentTunnelProvider,
    pub hostname: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpsertProjectPersistentHostnameInput {
    pub project_id: String,
    pub hostname: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyProjectPersistentHostnameInput {
    pub project_id: String,
    pub hostname: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistentTunnelSetupStatus {
    pub provider: PersistentTunnelProvider,
    pub ready: bool,
    pub managed: bool,
    pub binary_path: Option<String>,
    pub auth_cert_path: Option<String>,
    pub credentials_path: Option<String>,
    pub tunnel_id: Option<String>,
    pub tunnel_name: Option<String>,
    pub default_hostname_zone: Option<String>,
    pub details: String,
    pub guidance: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistentTunnelManagedSetup {
    pub provider: PersistentTunnelProvider,
    pub auth_cert_path: Option<String>,
    pub credentials_path: Option<String>,
    pub tunnel_id: Option<String>,
    pub tunnel_name: Option<String>,
    pub default_hostname_zone: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistentTunnelNamedTunnelSummary {
    pub tunnel_id: String,
    pub tunnel_name: String,
    pub credentials_path: Option<String>,
    pub selected: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePersistentNamedTunnelInput {
    pub name: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SelectPersistentNamedTunnelInput {
    pub tunnel_id: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdatePersistentTunnelSetupInput {
    pub default_hostname_zone: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PersistentTunnelStatus {
    Stopped,
    Starting,
    Running,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectPersistentTunnelState {
    pub project_id: String,
    pub provider: PersistentTunnelProvider,
    pub status: PersistentTunnelStatus,
    pub hostname: String,
    pub local_url: String,
    pub public_url: String,
    pub log_path: String,
    pub binary_path: Option<String>,
    pub tunnel_id: Option<String>,
    pub credentials_path: Option<String>,
    pub updated_at: String,
    pub details: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyProjectPersistentHostnameResult {
    pub hostname: ProjectPersistentHostname,
    pub tunnel: ProjectPersistentTunnelState,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteProjectPersistentHostnameResult {
    pub hostname: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistentTunnelHealthCheck {
    pub code: String,
    pub label: String,
    pub status: PersistentTunnelStatus,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistentTunnelHealthReport {
    pub project_id: String,
    pub hostname: Option<String>,
    pub overall_status: PersistentTunnelStatus,
    pub checks: Vec<PersistentTunnelHealthCheck>,
    pub updated_at: String,
}
